### ISSUE TITLE
Is there any way for podman to build a .sif container?

### ISSUE THREAD

**Issue by MaximilienLC** [containers/podman#18217]
It is currently possible to run .sif containers but is it possible to build them?

If not, is there any technique that would allow a similar smooth transition of:
build the .sif file -> transfer it to another machine -> run it directly without having to rebuild anything

**Comment by afbjorklund** [comment]
You can use `singularity build`, to convert an oci-archive directly into a sif file (generating the def).

`singularity build busybox.sif oci-archive://busybox.tar`

`podman run sif://$PWD/busybox.sif sh -c "echo Hello World"`

But it would probably be easier/better, to just run the `podman save` and `podman load` commands ?

* https://github.com/containers/podman/discussions/18007#discussioncomment-5506477

Using a registry allows for better and quicker workflow, as well...

**Comment by afbjorklund** [comment]
It would be _possible_ to add sif write support to containers/image (with siftool), but right now it is read-only.

https://github.com/containers/image/blob/v5.24.1/sif/transport.go#L154

Generating a **real** sif container image, does have some more strict requirements than docker/podman though.

https://apptainer.org/docs/user/main/docker_and_oci.html#differences-and-limitations-vs-docker

**Comment by MaximilienLC** [comment]
I'll try to be a little bit more precise about what I'm trying to do.

I want to run containers on several machines I do not have sudo access to. This includes SLURM clusters & regular machines.
The target area is ML and containers have the purpose of both training the systems (ex: on the SLURM cluster) and later deploy them (ex: on the non-sudo machines).

I have until now been runing a combo of Docker & Apptainer in order to achieve that goal (since Docker is not available on my SLURM cluster), but would like to replace that with a Podman only approach. My only issue with Podman in this setup is that I cannot simply build an image somewhere and transfer it to the SLURM cluster and execute it within a second, as I'm able to do with a .sif image. I've been trying various `podman save` options, with directories etc... to make it as quick as possible to `podman load` on the SLURM cluster and execute my training run, but the `podman load` always takes in the order of minutes to finish (the image size is ~15GB) which is a massive bottleneck in my training pipeline as it occurs several times.

Do you see any potential solution?

**Comment by afbjorklund** [comment]
I don't think the archive will scale to that type of image, but then again I don't understand what could possible be of that size - I would have guessed models and such (or databases, etc) would be on volumes, not baked into the image layers ?

So it seems like you are wanting some kind of shared image storage, and possibly even some smarter variant thereof. There was an article about different kind of remote solutions: https://www.redhat.com/sysadmin/image-stores-podman

**Comment by MaximilienLC** [comment]
haha the bulk of it is CUDA + PyTorch, there's no model nor database on it. Anyways thanks a lot for your help, I'll keep you posted if I find something reasonable.

**Comment by afbjorklund** [comment]
> My only issue with Podman in this setup is that I cannot simply build an image somewhere and transfer it to the SLURM cluster and execute it within a second, as I'm able to do with a .sif image.

I think that podman (et al) could benefit from being able to cache the import of .sif files, instead of always doing the unsquashfs and convertion to tar - which like you say could take forever for large images, and a while even for small

**Comment by afbjorklund** [comment]
Another way: https://www.redhat.com/sysadmin/podman-transfer-container-images-without-registry

**Comment by MaximilienLC** [comment]
So as I was mentioning earlier, one alternative would be to build/load the images on the login node, hoping that they would be available on the compute nodes. However, performing `podman load` on the compute node stores the image in my `$HOME` folder. Once the image is loaded and I run a job on the compute nodes however, podman becomes unfunctional, giving me this error anytime I try any command: `Error: mkdir /run/user/XXX: permission denied` (approximate error message, I'm rerunning it to get the exact one). I'm not sure if that would solve the problem but I've been trying to change the image location to the `/scratch/` filesystem is shared between both login & compute nodes. I've tried following several RedHat blogs (ex: https://www.redhat.com/sysadmin/rootless-podman-nfs) and changing the graphroot directory but podman then also becomes unusable on the login node, ex: `Error: mkdir /run/user/1000: permission denied`

**Comment by MaximilienLC** [comment]
> Another way: https://www.redhat.com/sysadmin/podman-transfer-container-images-without-registry

Nice, giving this a try

**Comment by MaximilienLC** [comment]
Thanks for all your help @afbjorklund , closing in favour of https://github.com/containers/podman/issues/18223

**Comment by afbjorklund** [comment]
Actually `podman image scp` just calls `podman save` and `podman load`, so it has the same issues...

If you want to live without a registry, you would have to check for existence and do the cache yourself.
That is, before starting the save/transfer/load process - check if the image is already present on remote.

I still think it would be a better idea to have a smarter registry, and use something like [IPFS / eStargz](https://archive.fosdem.org/2022/schedule/event/container_ipfs_image/) ?

**Comment by MaximilienLC** [comment]
I'm not opposed to using anything. Ultimately what I want is to be able to execute a container with an image available upon connecting to the compute node. Thus far anything loaded on the login node makes my whole podman unusable on the compute node (error creating tmpdir). The only way I've ever managed to run a container on the compute node has been to wipe clean the image storage and then build everything on the compute node.
